src/util/mpir_thread.c

```diff
@@ -0,0 +1,99 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2019 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ */
+
+#include <mpiimpl.h>
+
+#ifdef MPICH_IS_THREADED
+/* use MPIR_Add_mutex to register mutex, they will be created/destroyed together
+ * at init/finalize */
+
+#define MAX_MUTEX_COUNT 64
+static MPID_Thread_mutex_t *mutex_list[MAX_MUTEX_COUNT];
+static int mutex_count = 0;
+
+void MPIR_Add_mutex(MPID_Thread_mutex_t * p_mutex)
+{
+    MPIR_Assert(mutex_count < MAX_MUTEX_COUNT);
+    mutex_list[mutex_count++] = p_mutex;
+}
+
+/* All the other mutexes that depend on the configured thread granularity */
+/* FIXME: we currently have  various mutexes scattered throughout the code
+ * and their logic interweaved between thread model and objects/features.
+ * We should use a register mechanism to manage creation and destroying of
+ * all mutexes
+ */
+
+/* These routine handle any thread initialization that my be required */
+void MPIR_Thread_CS_Init(void)
```

Comment:
Shouldn't this have the `ENTER` and `EXIT` macros too? Similar comment for `Finalize`

Reply:
`CS_Init` initializes the mutexes, and `CS_Finalize` destroys the mutexes. They are not for entering/exiting critical sections.

Reply:
I was not referring to critical sections, rather `MPIR_FUNC_VERBOSE_ENTER/EXIT`. I am not sure myself when these should be used actually. We don't seem to be consistent in their use.

Reply:
Ahh... that, I'll add them.

Personally I really like to remove all of them :).

Reply:
On a second thought, I am not sure here since it is being called during `init` and adding these logging complicates the dependency. Let's deal with that separately from this PR.

Reply:
sounds good